pkg/controllers/deprovisioning/controller.go

```diff
@@ -411,57 +421,56 @@ func (c *Controller) waitForDeletion(ctx context.Context, node *v1.Node) {
 	}
 }
 
-// launchReplacementNode launches a replacement node and blocks until it is ready
-func (c *Controller) launchReplacementNode(ctx context.Context, action Command) error {
+// launchReplacementNodes launches replacement nodes and blocks until it is ready
+// nolint:gocyclo
+func (c *Controller) launchReplacementNodes(ctx context.Context, action Command) error {
 	defer metrics.Measure(deprovisioningReplacementNodeInitializedHistogram)()
-	if len(action.nodesToRemove) != 1 {
-		return fmt.Errorf("expected a single node to replace, found %d", len(action.nodesToRemove))
+	nodeNamesToRemove := lo.Map(action.nodesToRemove, func(n *v1.Node, _ int) string { return n.Name })
+	// cordon the old nodes before we launch the replacements to prevent new pods from scheduling to the old nodes
+	if err := c.setNodesUnschedulable(ctx, true, nodeNamesToRemove...); err != nil {
+		return fmt.Errorf("cordoning nodes, %w", err)
 	}
-	oldNode := action.nodesToRemove[0]
 
-	// cordon the node before we launch the replacement to prevent new pods from scheduling to the node
-	if err := c.setNodeUnschedulable(ctx, action.nodesToRemove[0].Name, true); err != nil {
-		return fmt.Errorf("cordoning node %s, %w", oldNode.Name, err)
-	}
-
-	nodeNames, err := c.provisioner.LaunchNodes(ctx, provisioning.LaunchOptions{RecordPodNomination: false}, action.replacementNode)
+	nodeNames, err := c.provisioner.LaunchNodes(ctx, provisioning.LaunchOptions{RecordPodNomination: false}, action.replacementNodes...)
 	if err != nil {
-		// uncordon the node as the launch may fail (e.g. ICE or incompatible AMI)
-		err = multierr.Append(err, c.setNodeUnschedulable(ctx, oldNode.Name, false))
+		// uncordon the nodes as the launch may fail (e.g. ICE or incompatible AMI)
+		err = multierr.Append(err, c.setNodesUnschedulable(ctx, false, nodeNamesToRemove...))
 		return err
 	}
-	if len(nodeNames) != 1 {
-		// shouldn't ever occur as we are only launching a single node
-		return fmt.Errorf("expected a single node name, got %d", len(nodeNames))
+	if len(nodeNames) != len(action.replacementNodes) {
```

Comment:
What happens if only some launches are successful?

Reply:
If only some launches are successful, we should still return an error, so this shouldn't be reached.

Reply:
If you're asking about expected behavior, in the context of Consolidation, I think it makes sense to not consolidate if we can't spawn all the nodes that we wanted to. In the context of Expiration, I think we could improve this since we want the node to be deleted anyway, but I think this is fine for now. WDYT?

Reply:
@njtran We should test this to discover what happens in a partial launch (like to due to ICE or some other error). It's not clear to me what the state would look like. At this point, we would have cordoned some nodes, we don't uncordon them on error (maybe we should?) and we would have more nodes on the cluster than we need.

Worth a sanity check I think.

Reply:
Will do.